### PR TITLE
exocom - rewrite in go: subsystems

### PR DESCRIPTION
**Pull request by charlierudolph**
made for @mcharawi

### CODE REVIEW

File: go/exocom/src/client_registry/client_registry.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package clientRegistry
+
+import "encoding/json"
+
+// Client is the combination of a client name, service type and internal namespace
+type Client struct {
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: Thoughts around extracting each struct into its own file with the struct name as the file name? That's how it is done in Ruby, and it's very helpful in finding implementations.

File: go/exocom/src/client_registry/client_registry.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	ClientName        string
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: Didn't we name this `role`? The word "name" is too generic. There are three names here:
- service type: the "class" of the service
- instance name (coming soon): the name of the particular service instance ("user service instance #3")
- role: the role that the service instance plays inside the application (a MongoService instance could play the role UserService). This is only different when using message translation.

File: go/exocom/src/client_registry/client_registry.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	ServiceType       string
+	InternalNamespace string
+}
+
+// Route is an entry in the routes
+type Route struct {
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: Shouldn't this be in some sort of subsystem, for example the broadcast subsystem?

File: go/exocom/src/client_registry/client_registry.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	Receives          []string
+	Sends             []string
+	InternalNamespace string
+}
+
+// ClientRegistry manages which clients are connected to exocom,
+// and what messages each client can send and receive
+type ClientRegistry struct {
+	Routing       map[string]Route
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: this line should be in the broadcast subsystem

File: go/exocom/src/client_registry/client_registry.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	Clients       map[string]Client
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: this map should be in the connection subsystem

File: go/exocom/src/client_registry/client_registry.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	subscriptions *SubscriptionManager
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry.go]: This one probably also goes into the broadcast subsystem. Maybe add convenience methods to add and look up entries to it as well?

File: go/exocom/src/client_registry/client_registry_test.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		BeforeEach(func() {
+			serviceRoutes :=
+				`{
+					"role 1": {
+						"receives": ["message 3 name"],
+						"internalNamespace": "namespace"
+					}
+				}`
+			var err error
+			registry, err = clientRegistry.NewClientRegistry(serviceRoutes)
```
Comment (kevgo) [inline go/exocom/src/client_registry/client_registry_test.go]: Why are we providing the routes as a string here? That seems strange. I would expect we parse config files into Go structs.

File: go/exocom/src/client_registry/subscription_manager.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// SubscriptionManager manages what clients should be sent messages
+type SubscriptionManager struct {
+	MessageNameToSubscribers map[string][]Subscriber
+	Routing                  map[string]Route
+}
+
+// NewSubscriptionManager returns a SubscriptionManager for the given routes
+func NewSubscriptionManager(routes map[string]Route) *SubscriptionManager {
```
Comment (kevgo) [inline go/exocom/src/client_registry/subscription_manager.go]: Can we defined a `Routes` type that is `map[string]Route` to make the semantics of data more explicit and not juggle so many anonymous maps around?

File: go/exocom/src/message_cache/message_cache.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	return
+}
+
+// Get returns the timestamp for the given messageId returning an error if
+// no data is available
+func (c *MessageCache) Get(messageID string) (time.Time, error) {
+	c.mutex.Lock()
+	timestamp, ok := c.Cache[messageID]
+	if ok {
+		c.mutex.Unlock()
```
Comment (kevgo) [inline go/exocom/src/message_cache/message_cache.go]: I think you can move this line before the if clause. It is the direct next action in both casese.

File: go/exocom/src/message_cache/message_cache.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+			delete(c.Cache, id)
+		}
+	}
+	return
+}
+
+// Get returns the timestamp for the given messageId returning an error if
+// no data is available
+func (c *MessageCache) Get(messageID string) (time.Time, error) {
+	c.mutex.Lock()
```
Comment (kevgo) [inline go/exocom/src/message_cache/message_cache.go]: Please add a comment why you are adding a lock here, it is not obvious from the code.

File: go/exocom/src/message_cache/message_cache.go (hunk 9 above)
Comment (charlierudolph) [inline go/exocom/src/message_cache/message_cache.go]: This got refactored in a later PR where the branch was removed

File: go/exocom/src/message_cache/message_cache.go (hunk 10 above)
Comment (charlierudolph) [inline go/exocom/src/message_cache/message_cache.go]: Sounds good. I'll do that in a separate PR

File: go/exocom/src/client_registry/subscription_manager.go (hunk 8 above)
Comment (charlierudolph) [inline go/exocom/src/client_registry/subscription_manager.go]: Sounds good

File: go/exocom/src/client_registry/client_registry_test.go (hunk 7 above)
Comment (charlierudolph) [inline go/exocom/src/client_registry/client_registry_test.go]: This is copying the node version of where the routes actually get parsed. We can move the parsing up the chain

File: go/exocom/src/client_registry/client_registry.go (hunk 6 above)
Comment (charlierudolph) [inline go/exocom/src/client_registry/client_registry.go]: I'd prefer to not add those until we need them

File: go/exocom/src/client_registry/client_registry.go (hunk 2 above)
Comment (charlierudolph) [inline go/exocom/src/client_registry/client_registry.go]: I believe its role in the service routes but was named clientName internally. Will update in a separate PR

File: go/exocom/src/client_registry/client_registry.go (hunk 3 above)
Comment (charlierudolph) [inline go/exocom/src/client_registry/client_registry.go]: For the rewrite we just kept the same internals as in the js version. Let explore the subsystems when you get back as I'm unclear what exactly lies within each boundary

**Pull request merged**